lib/js-yaml/dumper.js

```diff
@@ -150,81 +174,331 @@ function testImplicitResolving(state, str) {
+    er.position = position;
+    er.checkpoint = this.checkpoint;
+    throw er;
+  }
+
+  this.result += this.source.slice(this.checkpoint, position);
+  this.checkpoint = position;
+  return this;
+};
 
-  state.dump = '';
-  isQuoted = false;
-  checkpoint = 0;
-  first = object.charCodeAt(0) || 0;
+StringBuilder.prototype.escapeChar = function () {
+  var character = this.source.charCodeAt(this.checkpoint);
+  var esc = ESCAPE_SEQUENCES[character] || encodeHex(character);
+  this.result += esc;
+  this.checkpoint += 1;
+
+  return this;
+};
+
+StringBuilder.prototype.finish = function () {
+  if (this.source.length > this.checkpoint) {
+    this.takeUpTo(this.source.length);
+  }
+};
+
+
+function writeScalar (state, object, level) {
+  if (0 === object.length) {
+    state.dump = "''";
+    return;
+  }
 
   if (-1 !== DEPRECATED_BOOLEANS_SYNTAX.indexOf(object)) {
-    // Ensure compatibility with YAML 1.0/1.1 loaders.
-    isQuoted = true;
-  } else if (0 === object.length) {
-    // Quote empty string
-    isQuoted = true;
-  } else if (CHAR_SPACE    === first ||
-             CHAR_SPACE    === object.charCodeAt(object.length - 1)) {
-    isQuoted = true;
-  } else if (CHAR_MINUS    === first ||
-             CHAR_QUESTION === first) {
-    // Don't check second symbol for simplicity
-    isQuoted = true;
-  }
-
-  for (position = 0, length = object.length; position < length; position += 1) {
-    character = object.charCodeAt(position);
-
-    if (!isQuoted) {
-      if (CHAR_TAB                  === character ||
-          CHAR_LINE_FEED            === character ||
-          CHAR_CARRIAGE_RETURN      === character ||
-          CHAR_COMMA                === character ||
-          CHAR_LEFT_SQUARE_BRACKET  === character ||
-          CHAR_RIGHT_SQUARE_BRACKET === character ||
-          CHAR_LEFT_CURLY_BRACKET   === character ||
-          CHAR_RIGHT_CURLY_BRACKET  === character ||
-          CHAR_SHARP                === character ||
-          CHAR_AMPERSAND            === character ||
-          CHAR_ASTERISK             === character ||
-          CHAR_EXCLAMATION          === character ||
-          CHAR_VERTICAL_LINE        === character ||
-          CHAR_GREATER_THAN         === character ||
-          CHAR_SINGLE_QUOTE         === character ||
-          CHAR_DOUBLE_QUOTE         === character ||
-          CHAR_PERCENT              === character ||
-          CHAR_COMMERCIAL_AT        === character ||
-          CHAR_COLON                === character ||
-          CHAR_GRAVE_ACCENT         === character) {
-        isQuoted = true;
+    state.dump = "'" + object + "'";
+    return;
+  }
+
+  var simple = true;
+  var first = object.length ? object.charCodeAt(0) : 0;
+  var spaceWrap = (CHAR_SPACE === first ||
+                   CHAR_SPACE === object.charCodeAt(object.length - 1));
+
+  var maybeNum = (CHAR_MINUS    === first ||
+                  CHAR_QUESTION === first);
```

Comment:
Ah! Yeah, but it's actually not about numbers. Number-looking strings are secured by the implicit resolving check. It's actually a simplified check for "? " and "- " character sequences that denote mappings and sequences correspondingly. However there should be more of such characters, not only `-` and `?`. There is also `@` at least. See http://www.yaml.org/spec/1.2/spec.html#ns-plain-first%28c%29